### PR TITLE
AMDGPU: Define v_mfma_f32_{16x16x128|32x32x64}_f8f6f4 instructions

### PR DESCRIPTION
**Pull request by arsenm**
These use a new VOP3PX encoding for the v_mfma_scale_* instructions,
which bundles the pre-scale v_mfma_ld_scale_b32. None of the modifiers
are supported yet (op_sel, neg or clamp).

I'm not sure the intrinsic should really expose op_sel (or any of the
others). If I'm reading the documentation correctly, we should be able
to just have the raw scale operands and auto-match op_sel to byte
extract patterns.

The op_sel syntax also seems extra horrible in this usage, especially with the
usual assumed op_sel_hi=-1 behavior.

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116723?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#116724** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116724?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116723** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116723?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116723?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#116722** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116722?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116681** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116681?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116680** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116680?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116679** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116679?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116678** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116678?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116312** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116312?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116311** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116311?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116310** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116310?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116309** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116309?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116308** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116308?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116307** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116307?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: llvm/docs/AMDGPUUsage.rst [hunk 1]
```diff
@@ -1397,6 +1397,19 @@ The AMDGPU backend implements the following LLVM IR intrinsics.
                                                    used by hardware to control active lanes when used in EXEC register.
                                                    For example, ballot(i1 true) return EXEC mask.
 
+  llvm.amdgcn.mfma.f32.16x16x128.f8f6f4.scaled     Emit `v_mfma_f32_16x16x128_f8f6f4`, bundled with a `v_mfma_ld_scale_b32`
```
Comment (shiltian) [inline llvm/docs/AMDGPUUsage.rst:1400]: This reminds me that we probably didn't add other gfx950 intrinsics to the document.

**Review by arsenm — COMMENTED** [review]
Going to put this one off to reduce future conflicts

**Review by shiltian — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -15454,6 +15454,23 @@ void SITargetLowering::AdjustInstrPostInstrSelection(MachineInstr &MI,
...
+      if (TII->isMAI(MI)) {
+        // The ordinary src0, src1, src2 were legalized above.
+        //
+        // We have to also legalize the appended v_mfma_ld_scale_b32 operands,
+        // as a separate instruction.
+        int Src0Idx = AMDGPU::getNamedOperandIdx(MI.getOpcode(),
+                                                 AMDGPU::OpName::scale_src0);
+        if (Src0Idx != -1) {
+          int Src1Idx = Src0Idx + 2;
+          assert(Src1Idx = AMDGPU::getNamedOperandIdx(
```
Comment (mariusz-sikora-at-amd) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:15466]: == ?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:15466]: yes

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:15466]: Fixed by #117238

**Pull request merged**